// dora-plugin-livereload/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _fs = require('fs');

var _url = require('url');

var _path = require('path');

var _lodash = require('lodash.isequal');

var _lodash2 = _interopRequireDefault(_lodash);

var _tinyLr = require('tiny-lr');

var _tinyLr2 = _interopRequireDefault(_tinyLr);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var localIP = require('internal-ip')();

var lrOpts = {};

var ignoreOpts = {
  enableJs: true,
  enableCss: true,
  enableImg: true,
  enableAll: false
};
var ignorePattern = {
  enableJs: 'js',
  enableCss: 'css',
  enableImg: 'jpg|jpeg|gif|png|bmp'
};
var pattern = '';
var tinylrServer = {};
var firstRun = 0;
var preCompilerationAssets = {};

function getAssetContent(asset) {
  var content = null;
  if (asset._value) {
    content = asset._value;
  } else if (asset.children) {
    content = asset.children;
  } else if (asset._cachedSource) {
    content = asset._cachedSource;
  }

  return content;
}

function getPattern(opts) {
  var patternString = Object.keys(opts).reduce(function (prev, item) {
    if (opts[item]) {
      prev.push(ignorePattern[item]);
    }

    return prev;
  }, []);

  return patternString.join('|');
}

exports.default = {
  'middleware.before': function middlewareBefore() {
    var log = this.log;
    var query = this.query;

    if (query && (typeof query === 'undefined' ? 'undefined' : _typeof(query)) === 'object') {
      ignoreOpts = _extends({}, ignoreOpts, query);
      if (ignoreOpts.enableAll) {
        pattern = '.*$';
      } else {
        pattern = '.(' + getPattern(ignoreOpts) + ')$';
      }
    }
    lrOpts = {
      port: 35729,
      errorListener: function errorListener(err) {
        log.error(err);
      }
    };
    tinylrServer = (0, _tinyLr2.default)(lrOpts);
    tinylrServer.listen(lrOpts.port, function () {
      log.info('listening on ' + lrOpts.port);
    });
  },
  'middleware': function middleware() {
    var cwd = this.cwd;
    var log = this.log;

    var isNeedLiveReload = true;
    var reg = void 0;
    if (pattern.length !== 1) {
      reg = new RegExp(pattern, 'i');
      log.debug('livereload is watching the pattern of ' + pattern + ' files');
    } else {
      isNeedLiveReload = false;
    }

    var compiler = this.get('compiler');
    if (!compiler) {
      throw new Error('[error] must used together with dora-plugin-atool-build');
    }
    compiler.plugin('done', function doneHandler(stats) {
      if (stats.hasErrors()) {
        log.error(stats.toString());

        return;
      }

      var assets = stats.compilation.assets;
      if (!isNeedLiveReload) {
        return;
      }

      var items = [];
      items = Object.keys(assets).filter(function (item) {
        return reg.test(item) && (0, _path.extname)(item) !== '.map';
      });
      log.debug('final watching items ' + items);
      if (!firstRun) {
        firstRun++;
        preCompilerationAssets = items.reduce(function (prev, item) {
          prev[item] = getAssetContent(assets[item]);

          return prev;
        }, {});

        return;
      }

      var changedItems = items.reduce(function (prev, item) {
        var pre = preCompilerationAssets[item];
        var now = getAssetContent(assets[item]);

        if (!(0, _lodash2.default)(pre, now)) {
          preCompilerationAssets[item] = now;
          prev.push(item);
        }

        return prev;
      }, []);

      tinylrServer.changed({ body: { files: changedItems } });
      log.info('livereload changed ' + changedItems.join(', '));
    });

    return regeneratorRuntime.mark(function _callee(next) {
      var pathName, fileName, filePath, isHTML, injectScript, content, docTypeReg, docType;
      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              pathName = (0, _url.parse)(this.url).pathname;
              fileName = pathName === '/' ? 'index.html' : pathName;
              filePath = (0, _path.join)(cwd, fileName);
              isHTML = /\.html?$/.test(fileName);

              if (!(isHTML && (0, _fs.existsSync)(filePath))) {
                _context.next = 16;
                break;
              }

              injectScript = '<script src=\'http://' + localIP + ':' + lrOpts.port + '/livereload.js\'></script>';
              content = (0, _fs.readFileSync)(filePath, 'utf-8');
              docTypeReg = new RegExp('^\s*\<\!DOCTYPE\s*.+\>.*$', 'im');
              docType = content.match(docTypeReg);

              if (!docType) {
                _context.next = 13;
                break;
              }

              content = content.replace(docTypeReg, docType[0] + injectScript);
              this.body = content;

              return _context.abrupt('return');

            case 13:
              content = injectScript + content;
              this.body = content;

              return _context.abrupt('return');

            case 16:
              _context.next = 18;
              return next;

            case 18:
            case 'end':
              return _context.stop();
          }
        }
      }, _callee, this);
    });
  }
};
module.exports = exports['default'];
